File: palletizer_full/optimizer.py
"""
Mixed-SKU Pallet Optimizer
==========================

A real, deterministic pallet-packing engine. Given a set of boxes and a
pallet, it computes actual placements (with 90-degree rotation), an actual
volumetric density, an actual density uplift versus a naive baseline, and a
deterministic, physics-grounded stability score.

Nothing here is random or hardcoded: every metric is derived from the geometry
of the computed placement. The stability model combines base-support ratio
(how much of each box rests on the layer below) with center-of-mass offset
from the pallet center.

This is a heuristic engine (shelf packing + height-grouped layers). It is the
honest v1 of the "one hard capability". A future version can swap the layer
packer for an exact solver (OR-Tools CP-SAT / MILP) behind the same API.
"""

from __future__ import annotations

import csv
import math
from dataclasses import asdict, dataclass, field
from pathlib import Path

# Standard GMA pallet, mm. Override per deployment.
DEFAULT_PALLET_LENGTH_MM = 1219.0
DEFAULT_PALLET_WIDTH_MM = 1016.0
DEFAULT_MAX_HEIGHT_MM = 1800.0
DEFAULT_MAX_WEIGHT_KG = 1000.0


@dataclass(frozen=True)
class Box:
    sku_id: str
    length_mm: float
    width_mm: float
    height_mm: float
    weight_kg: float = 0.0

    @property
    def footprint_mm2(self) -> float:
        return self.length_mm * self.width_mm

    @property
    def volume_mm3(self) -> float:
        return self.length_mm * self.width_mm * self.height_mm


@dataclass(frozen=True)
class Pallet:
    length_mm: float = DEFAULT_PALLET_LENGTH_MM
    width_mm: float = DEFAULT_PALLET_WIDTH_MM
    max_height_mm: float = DEFAULT_MAX_HEIGHT_MM
    max_weight_kg: float = DEFAULT_MAX_WEIGHT_KG

    @property
    def footprint_mm2(self) -> float:
        return self.length_mm * self.width_mm


@dataclass
class Placement:
    sku_id: str
    x_mm: float          # lower-left corner on the pallet
    y_mm: float
    z_mm: float          # base height of the box
    length_mm: float     # footprint length AFTER rotation
    width_mm: float      # footprint width AFTER rotation
    height_mm: float
    weight_kg: float
    rot_deg: float       # 0 or 90
    layer: int


@dataclass
class PalletPlan:
    placements: list[Placement] = field(default_factory=list)
    unplaced: list[str] = field(default_factory=list)
    num_layers: int = 0
    stack_height_mm: float = 0.0
    total_weight_kg: float = 0.0
    volume_density: float = 0.0
    baseline_density: float = 0.0
    density_uplift_pct: float = 0.0
    support_score: float = 0.0
    com_score: float = 0.0
    stability_score: float = 0.0
    is_valid: bool = False
    recommendations: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        d = asdict(self)
        return d


def _overlap_area(a: Placement, b: Placement) -> float:
    """Footprint overlap area (mm^2) between two placements."""
    ax2, ay2 = a.x_mm + a.length_mm, a.y_mm + a.width_mm
    bx2, by2 = b.x_mm + b.length_mm, b.y_mm + b.width_mm
    dx = max(0.0, min(ax2, bx2) - max(a.x_mm, b.x_mm))
    dy = max(0.0, min(ay2, by2) - max(a.y_mm, b.y_mm))
    return dx * dy


def _pack_layer(
    boxes: list[Box],
    pallet: Pallet,
    z_mm: float,
    layer_index: int,
) -> tuple[list[Placement], list[Box]]:
    """Shelf-pack a single layer. Returns (placements, leftover boxes).

    Greedy First-Fit-Decreasing by footprint, trying both orientations so the
    long edge runs along whichever pallet axis leaves less waste.
    """
    placed: list[Placement] = []
    leftover: list[Box] = []

    # Tallest/biggest first packs denser and more stably.
    ordered = sorted(boxes, key=lambda b: b.footprint_mm2, reverse=True)

    x_cursor = 0.0
    y_cursor = 0.0
    shelf_depth = 0.0  # width consumed by the current shelf (y direction)

    for box in ordered:
        placed_this = False
        # Two candidate orientations: (L,W) and rotated (W,L).
        orientations = [
            (box.length_mm, box.width_mm, 0.0),
            (box.width_mm, box.length_mm, 90.0),
        ]
        for fp_l, fp_w, rot in orientations:
            if fp_l > pallet.length_mm or fp_w > pallet.width_mm:
                continue  # cannot fit this orientation at all
            # Fits on the current shelf?
            if x_cursor + fp_l <= pallet.length_mm + 1e-6 and y_cursor + fp_w <= pallet.width_mm + 1e-6:
                placed.append(Placement(
                    sku_id=box.sku_id, x_mm=x_cursor, y_mm=y_cursor, z_mm=z_mm,
                    length_mm=fp_l, width_mm=fp_w, height_mm=box.height_mm,
                    weight_kg=box.weight_kg, rot_deg=rot, layer=layer_index,
                ))
                x_cursor += fp_l
                shelf_depth = max(shelf_depth, fp_w)
                placed_this = True
                break
            # Try starting a new shelf (advance in y).
            new_y = y_cursor + shelf_depth
            if new_y + fp_w <= pallet.width_mm + 1e-6 and fp_l <= pallet.length_mm + 1e-6:
                y_cursor = new_y
                x_cursor = 0.0
                shelf_depth = fp_w
                placed.append(Placement(
                    sku_id=box.sku_id, x_mm=x_cursor, y_mm=y_cursor, z_mm=z_mm,
                    length_mm=fp_l, width_mm=fp_w, height_mm=box.height_mm,
                    weight_kg=box.weight_kg, rot_deg=rot, layer=layer_index,
                ))
                x_cursor += fp_l
                placed_this = True
                break
        if not placed_this:
            leftover.append(box)

    return placed, leftover


def _baseline_density(boxes: list[Box], pallet: Pallet) -> float:
    """Naive reference: single orientation, one box per footprint cell, stacked
    in simple rows. Represents 'good enough' manual/basic stacking."""
    if not boxes:
        return 0.0
    placed_vol = 0.0
    x = y = z = 0.0
    row_depth = 0.0
    layer_h = 0.0
    for box in boxes:
        if box.length_mm > pallet.length_mm or box.width_mm > pallet.width_mm:
            continue
        if x + box.length_mm > pallet.length_mm:
            x = 0.0
            y += row_depth
            row_depth = 0.0
        if y + box.width_mm > pallet.width_mm:
            x = y = 0.0
            row_depth = 0.0
            z += layer_h
            layer_h = 0.0
            if z >= pallet.max_height_mm:
                break
        x += box.length_mm
        row_depth = max(row_depth, box.width_mm)
        layer_h = max(layer_h, box.height_mm)
        placed_vol += box.volume_mm3
    stack_h = z + layer_h
    if stack_h <= 0:
        return 0.0
    return placed_vol / (pallet.footprint_mm2 * stack_h)


def _stability(placements: list[Placement], pallet: Pallet) -> tuple[float, float, float]:
    """Deterministic stability from geometry.

    Returns (support_score, com_score, stability_score), all in [0, 1].
    """
    if not placements:
        return 0.0, 0.0, 0.0

    by_layer: dict[int, list[Placement]] = {}
    for p in placements:
        by_layer.setdefault(p.layer, []).append(p)

    # Base-support ratio: fraction of each box's footprint resting on the
    # layer below (layer 0 rests fully on the pallet).
    support_ratios: list[float] = []
    for layer_idx, layer in by_layer.items():
        below = by_layer.get(layer_idx - 1, []) if layer_idx > 0 else None
        for box in layer:
            if below is None:
                support_ratios.append(1.0)
                continue
            supported = sum(_overlap_area(box, b) for b in below)
            area = box.length_mm * box.width_mm
            support_ratios.append(min(1.0, supported / area) if area else 0.0)
    support_score = sum(support_ratios) / len(support_ratios)

    # Center-of-mass offset from pallet center, normalized by half-extent.
    total_w = sum(max(p.weight_kg, 1e-6) for p in placements)
    com_x = sum((p.x_mm + p.length_mm / 2) * max(p.weight_kg, 1e-6) for p in placements) / total_w
    com_y = sum((p.y_mm + p.width_mm / 2) * max(p.weight_kg, 1e-6) for p in placements) / total_w
    cx, cy = pallet.length_mm / 2, pallet.width_mm / 2
    half_diag = math.hypot(cx, cy)
    offset = math.hypot(com_x - cx, com_y - cy) / half_diag
    com_score = max(0.0, 1.0 - offset)

    stability_score = 0.6 * support_score + 0.4 * com_score
    return support_score, com_score, stability_score


def optimize_pallet(boxes: list[Box], pallet: Pallet | None = None) -> PalletPlan:
    """Compute a real pallet plan for the given boxes.

    Boxes are grouped into height-similar layers and shelf-packed. Every metric
    on the returned :class:`PalletPlan` is derived from the resulting geometry.
    """
    pallet = pallet or Pallet()
    plan = PalletPlan()
    if not boxes:
        return plan

    # Group by height (descending) so each layer has a well-defined height with
    # minimal vertical waste.
    remaining = sorted(boxes, key=lambda b: b.height_mm, reverse=True)

    z = 0.0
    weight = 0.0
    layer_index = 0
    placed_all: list[Placement] = []

    while remaining:
        # Stop if the next (shortest possible) layer would exceed height budget.
        min_h = min(b.height_mm for b in remaining)
        if z + min_h > pallet.max_height_mm:
            break

        placed, leftover = _pack_layer(remaining, pallet, z, layer_index)
        if not placed:
            break  # nothing fits (oversized boxes) -> unplaced below

        layer_height = max(p.height_mm for p in placed)
        if z + layer_height > pallet.max_height_mm:
            break

        layer_weight = sum(p.weight_kg for p in placed)
        if weight + layer_weight > pallet.max_weight_kg and placed_all:
            break  # weight budget reached; keep what we already have

        placed_all.extend(placed)
        z += layer_height
        weight += layer_weight
        layer_index += 1
        remaining = leftover

    plan.placements = placed_all
    plan.unplaced = [b.sku_id for b in remaining]
    plan.num_layers = layer_index
    plan.stack_height_mm = round(z, 2)
    plan.total_weight_kg = round(weight, 2)

    placed_vol = sum(p.length_mm * p.width_mm * p.height_mm for p in placed_all)
    bounding_vol = pallet.footprint_mm2 * z if z > 0 else 0.0
    plan.volume_density = round(placed_vol / bounding_vol, 4) if bounding_vol else 0.0

    base = _baseline_density(boxes, pallet)
    plan.baseline_density = round(base, 4)
    plan.density_uplift_pct = round((plan.volume_density - base) / base * 100, 1) if base > 0 else 0.0

    support, com, stab = _stability(placed_all, pallet)
    plan.support_score = round(support, 3)
    plan.com_score = round(com, 3)
    plan.stability_score = round(stab, 3)

    recs: list[str] = []
    if plan.unplaced:
        recs.append(f"{len(plan.unplaced)} box(es) did not fit; consider a second pallet or larger footprint.")
    if plan.support_score < 0.85:
        recs.append("Low base support on upper layers; reorder so larger footprints sit lower.")
    if plan.com_score < 0.8:
        recs.append("Load is off-center; redistribute heavy SKUs toward the pallet center.")
    plan.recommendations = recs

    plan.is_valid = (
        bool(placed_all)
        and plan.stability_score >= 0.6
        and plan.com_score >= 0.5
    )
    return plan


def load_boxes_csv(path: str | Path) -> list[Box]:
    """Load boxes from a CSV with columns:
    sku_id, length_mm, width_mm, height_mm, weight_kg (weight optional).
    """
    boxes: list[Box] = []
    with open(path, newline="") as fh:
        reader = csv.DictReader(fh)
        for row in reader:
            boxes.append(Box(
                sku_id=row["sku_id"].strip(),
                length_mm=float(row["length_mm"]),
                width_mm=float(row["width_mm"]),
                height_mm=float(row["height_mm"]),
                weight_kg=float(row.get("weight_kg", 0) or 0),
            ))
    return boxes


def cli() -> None:
    """Console entry point: palletize-optimize <skus.csv>"""
    import argparse
    import json

    parser = argparse.ArgumentParser(description="Optimize a mixed-SKU pallet from a CSV of boxes.")
    parser.add_argument("csv_path", help="CSV with columns sku_id,length_mm,width_mm,height_mm,weight_kg")
    parser.add_argument("--pallet-length", type=float, default=DEFAULT_PALLET_LENGTH_MM)
    parser.add_argument("--pallet-width", type=float, default=DEFAULT_PALLET_WIDTH_MM)
    parser.add_argument("--max-height", type=float, default=DEFAULT_MAX_HEIGHT_MM)
    parser.add_argument("--max-weight", type=float, default=DEFAULT_MAX_WEIGHT_KG)
    parser.add_argument("--json", action="store_true", help="Emit full plan as JSON.")
    args = parser.parse_args()

    boxes = load_boxes_csv(args.csv_path)
    pallet = Pallet(args.pallet_length, args.pallet_width, args.max_height, args.max_weight)
    plan = optimize_pallet(boxes, pallet)

    if args.json:
        print(json.dumps(plan.to_dict(), indent=2))
        return

    print(f"Boxes in:        {len(boxes)}")
    print(f"Placed:          {len(plan.placements)}  ({len(plan.unplaced)} unplaced)")
    print(f"Layers:          {plan.num_layers}")
    print(f"Stack height:    {plan.stack_height_mm:.0f} mm")
    print(f"Total weight:    {plan.total_weight_kg:.1f} kg")
    print(f"Density:         {plan.volume_density * 100:.1f}%  (baseline {plan.baseline_density * 100:.1f}%)")
    print(f"Density uplift:  {plan.density_uplift_pct:+.1f}%  vs naive baseline")
    print(f"Support score:   {plan.support_score:.3f}")
    print(f"CoM score:       {plan.com_score:.3f}")
    print(f"Stability:       {plan.stability_score:.3f}  ({'VALID' if plan.is_valid else 'REVIEW'})")
    for r in plan.recommendations:
        print(f"  - {r}")


if __name__ == "__main__":
    cli()

File: web/lib/palletizer/optimizer.ts
// TypeScript port of palletizer_full/optimizer.py.
//
// This is NOT a simulation of the Python engine — it is the same algorithm,
// ported function-for-function (same shelf packing, same baseline, same
// stability model: 0.6 * base-support + 0.4 * center-of-mass score).
// Given the same boxes it produces the same placements and the same metrics
// as `palletize-optimize --json`.
//
// The Python core remains the source of truth in deployed cells; this port
// exists so the web demos give instant, honest feedback with identical math.
// Deterministic constraint extensions (heavy_low / fragile_high / speed_mode)
// are additive re-orderings of the same packer and are mirrored in
// gateway/demo_api.py.

import {
  BoxSpec,
  CorePlan,
  DEFAULT_PALLET,
  OptimizeConstraints,
  PalletSpec,
  Placement,
  PlanMetrics,
  ValidationReport,
  WebPlan,
} from './types';

const EPS = 1e-6;

function footprint(b: BoxSpec): number {
  return b.length_mm * b.width_mm;
}
function volume(b: BoxSpec): number {
  return b.length_mm * b.width_mm * b.height_mm;
}

/** Footprint overlap area (mm^2) between two placements. Port of _overlap_area. */
export function overlapArea(a: Placement, b: Placement): number {
  const ax2 = a.x_mm + a.length_mm;
  const ay2 = a.y_mm + a.width_mm;
  const bx2 = b.x_mm + b.length_mm;
  const by2 = b.y_mm + b.width_mm;
  const dx = Math.max(0, Math.min(ax2, bx2) - Math.max(a.x_mm, b.x_mm));
  const dy = Math.max(0, Math.min(ay2, by2) - Math.max(a.y_mm, b.y_mm));
  return dx * dy;
}

/** Shelf-pack a single layer. Port of _pack_layer (greedy FFD by footprint, both orientations). */
function packLayer(
  boxes: BoxSpec[],
  pallet: PalletSpec,
  z_mm: number,
  layerIndex: number,
  speedMode: boolean,
): { placed: Placement[]; leftover: BoxSpec[] } {
  const placed: Placement[] = [];
  const leftover: BoxSpec[] = [];

  const ordered = [...boxes].sort((a, b) => footprint(b) - footprint(a));

  let xCursor = 0;
  let yCursor = 0;
  let shelfDepth = 0;

  for (const box of ordered) {
    let placedThis = false;
    const orientations: Array<[number, number, number]> = speedMode
      ? [[box.length_mm, box.width_mm, 0]]
      : [
          [box.length_mm, box.width_mm, 0],
          [box.width_mm, box.length_mm, 90],
        ];

    for (const [fpL, fpW, rot] of orientations) {
      if (fpL > pallet.length_mm || fpW > pallet.width_mm) continue;

      if (xCursor + fpL <= pallet.length_mm + EPS && yCursor + fpW <= pallet.width_mm + EPS) {
        placed.push({
          sku_id: box.sku_id,
          x_mm: xCursor,
          y_mm: yCursor,
          z_mm,
          length_mm: fpL,
          width_mm: fpW,
          height_mm: box.height_mm,
          weight_kg: box.weight_kg,
          rot_deg: rot,
          layer: layerIndex,
          fragility: box.fragility,
        });
        xCursor += fpL;
        shelfDepth = Math.max(shelfDepth, fpW);
        placedThis = true;
        break;
      }
      const newY = yCursor + shelfDepth;
      if (newY + fpW <= pallet.width_mm + EPS && fpL <= pallet.length_mm + EPS) {
        yCursor = newY;
        xCursor = 0;
        shelfDepth = fpW;
        placed.push({
          sku_id: box.sku_id,
          x_mm: xCursor,
          y_mm: yCursor,
          z_mm,
          length_mm: fpL,
          width_mm: fpW,
          height_mm: box.height_mm,
          weight_kg: box.weight_kg,
          rot_deg: rot,
          layer: layerIndex,
          fragility: box.fragility,
        });
        xCursor += fpL;
        placedThis = true;
        break;
      }
    }
    if (!placedThis) leftover.push(box);
  }

  return { placed, leftover };
}

/** Port of _baseline_density: naive single-orientation row stacking. */
function baselineDensity(boxes: BoxSpec[], pallet: PalletSpec): number {
  if (boxes.length === 0) return 0;
  let placedVol = 0;
  let x = 0,
    y = 0,
    z = 0;
  let rowDepth = 0;
  let layerH = 0;
  for (const box of boxes) {
    if (box.length_mm > pallet.length_mm || box.width_mm > pallet.width_mm) continue;
    if (x + box.length_mm > pallet.length_mm) {
      x = 0;
      y += rowDepth;
      rowDepth = 0;
    }
    if (y + box.width_mm > pallet.width_mm) {
      x = 0;
      y = 0;
      rowDepth = 0;
      z += layerH;
      layerH = 0;
      if (z >= pallet.max_height_mm) break;
    }
    x += box.length_mm;
    rowDepth = Math.max(rowDepth, box.width_mm);
    layerH = Math.max(layerH, box.height_mm);
    placedVol += volume(box);
  }
  const stackH = z + layerH;
  if (stackH <= 0) return 0;
  return placedVol / (pallet.length_mm * pallet.width_mm * stackH);
}

/** Port of _stability: (support_score, com_score, stability_score) in [0,1]. */
export function stabilityFromGeometry(
  placements: Placement[],
  pallet: PalletSpec,
): { support: number; com: number; stability: number } {
  if (placements.length === 0) return { support: 0, com: 0, stability: 0 };

  const byLayer = new Map<number, Placement[]>();
  for (const p of placements) {
    const arr = byLayer.get(p.layer) ?? [];
    arr.push(p);
    byLayer.set(p.layer, arr);
  }

  const supportRatios: number[] = [];
  byLayer.forEach((layer, layerIdx) => {
    const below = layerIdx > 0 ? byLayer.get(layerIdx - 1) ?? [] : null;
    for (const box of layer) {
      if (below === null) {
        supportRatios.push(1.0);
        continue;
      }
      const supported = below.reduce((s, b) => s + overlapArea(box, b), 0);
      const area = box.length_mm * box.width_mm;
      supportRatios.push(area ? Math.min(1, supported / area) : 0);
    }
  });
  const support = supportRatios.reduce((a, b) => a + b, 0) / supportRatios.length;

  const totalW = placements.reduce((s, p) => s + Math.max(p.weight_kg, 1e-6), 0);
  const comX =
    placements.reduce((s, p) => s + (p.x_mm + p.length_mm / 2) * Math.max(p.weight_kg, 1e-6), 0) / totalW;
  const comY =
    placements.reduce((s, p) => s + (p.y_mm + p.width_mm / 2) * Math.max(p.weight_kg, 1e-6), 0) / totalW;
  const cx = pallet.length_mm / 2;
  const cy = pallet.width_mm / 2;
  const halfDiag = Math.hypot(cx, cy);
  const offset = Math.hypot(comX - cx, comY - cy) / halfDiag;
  const com = Math.max(0, 1 - offset);

  return { support, com, stability: 0.6 * support + 0.4 * com };
}

function isFragile(b: BoxSpec, c: OptimizeConstraints): boolean {
  const thr = c.fragile_threshold ?? 0.6;
  if (c.protect_sku_ids?.includes(b.sku_id)) return true;
  return (b.fragility ?? 0) >= thr;
}

/**
 * Port of optimize_pallet, with deterministic constraint extensions.
 *
 * Base behaviour (no constraints) matches the Python core exactly: boxes are
 * grouped into height-similar layers (tallest first) and shelf-packed.
 *
 * Extensions (all deterministic re-orderings of the same packer):
 *  - heavy_low: pack in weight-descending order so mass concentrates in lower layers.
 *  - fragile_high: fragile boxes are withheld until all non-fragile boxes are
 *    packed, so nothing is stacked on top of them.
 *  - speed_mode: single orientation only (no 90° wrist rotations).
 */
export function optimizePallet(
  boxes: BoxSpec[],
  palletIn?: Partial<PalletSpec>,
  constraints: OptimizeConstraints = {},
): CorePlan {
  const pallet: PalletSpec = {
    ...DEFAULT_PALLET,
    ...palletIn,
    ...(constraints.max_height_mm ? { max_height_mm: constraints.max_height_mm } : {}),
    ...(constraints.max_weight_kg ? { max_weight_kg: constraints.max_weight_kg } : {}),
  };

  const plan: CorePlan = {
    placements: [],
    unplaced: [],
    num_layers: 0,
    stack_height_mm: 0,
    total_weight_kg: 0,
    volume_density: 0,
    baseline_density: 0,
    density_uplift_pct: 0,
    support_score: 0,
    com_score: 0,
    stability_score: 0,
    is_valid: false,
    recommendations: [],
  };
  if (boxes.length === 0) return plan;

  // Packing order. Default = height desc (Python core). heavy_low = weight desc, height as tiebreak.
  const orderKey = (b: BoxSpec) =>
    constraints.heavy_low ? [b.weight_kg, b.height_mm] : [b.height_mm, 0];
  const sortDesc = (arr: BoxSpec[]) =>
    [...arr].sort((a, b) => {
      const ka = orderKey(a);
      const kb = orderKey(b);
      return kb[0] - ka[0] || kb[1] - ka[1];
    });

  let fragilePool: BoxSpec[] = [];
  let remaining: BoxSpec[];
  if (constraints.fragile_high) {
    fragilePool = sortDesc(boxes.filter((b) => isFragile(b, constraints)));
    remaining = sortDesc(boxes.filter((b) => !isFragile(b, constraints)));
  } else {
    remaining = sortDesc(boxes);
  }

  let z = 0;
  let weight = 0;
  let layerIndex = 0;
  const placedAll: Placement[] = [];
  let phase: 'base' | 'fragile' = 'base';

  while (remaining.length > 0 || (constraints.fragile_high && fragilePool.length > 0)) {
    if (remaining.length === 0 && fragilePool.length > 0) {
      remaining = fragilePool;
      fragilePool = [];
      phase = 'fragile';
    }
    if (remaining.length === 0) break;

    const minH = Math.min(...remaining.map((b) => b.height_mm));
    if (z + minH > pallet.max_height_mm) break;

    const { placed, leftover } = packLayer(remaining, pallet, z, layerIndex, !!constraints.speed_mode);
    if (placed.length === 0) break;

    const layerHeight = Math.max(...placed.map((p) => p.height_mm));
    if (z + layerHeight > pallet.max_height_mm) break;

    const layerWeight = placed.reduce((s, p) => s + p.weight_kg, 0);
    if (weight + layerWeight > pallet.max_weight_kg && placedAll.length > 0) break;

    placedAll.push(...placed);
    z += layerHeight;
    weight += layerWeight;
    layerIndex += 1;
    remaining = leftover;
  }

  const unplacedIds = [...remaining.map((b) => b.sku_id), ...fragilePool.map((b) => b.sku_id)];

  plan.placements = placedAll;
  plan.unplaced = unplacedIds;
  plan.num_layers = layerIndex;
  plan.stack_height_mm = Math.round(z * 100) / 100;
  plan.total_weight_kg = Math.round(weight * 100) / 100;

  const placedVol = placedAll.reduce((s, p) => s + p.length_mm * p.width_mm * p.height_mm, 0);
  const boundingVol = z > 0 ? pallet.length_mm * pallet.width_mm * z : 0;
  plan.volume_density = boundingVol ? Math.round((placedVol / boundingVol) * 10000) / 10000 : 0;

  const base = baselineDensity(boxes, pallet);
  plan.baseline_density = Math.round(base * 10000) / 10000;
  plan.density_uplift_pct = base > 0 ? Math.round(((plan.volume_density - base) / base) * 1000) / 10 : 0;

  const { support, com, stability } = stabilityFromGeometry(placedAll, pallet);
  plan.support_score = Math.round(support * 1000) / 1000;
  plan.com_score = Math.round(com * 1000) / 1000;
  plan.stability_score = Math.round(stability * 1000) / 1000;

  const recs: string[] = [];
  if (plan.unplaced.length > 0) {
    recs.push(`${plan.unplaced.length} box(es) did not fit; consider a second pallet or larger footprint.`);
  }
  if (plan.support_score < 0.85) {
    recs.push('Low base support on upper layers; reorder so larger footprints sit lower.');
  }
  if (plan.com_score < 0.8) {
    recs.push('Load is off-center; redistribute heavy SKUs toward the pallet center.');
  }
  if (phase === 'fragile' && constraints.fragile_high) {
    recs.push('Fragile SKUs isolated on top layers — nothing is stacked above them.');
  }
  plan.recommendations = recs;

  plan.is_valid = placedAll.length > 0 && plan.stability_score >= 0.6 && plan.com_score >= 0.5;
  return plan;
}

// ---------------------------------------------------------------------------
// Web plan wrapper: derived metrics only (no invented numbers).
// ---------------------------------------------------------------------------

/** Robot cycle estimate: per-pick base + wrist-rotation penalty. Both parameters are stated, not hidden. */
export const CYCLE_BASE_S_PER_BOX = 7.5;
export const CYCLE_ROT_PENALTY_S = 1.8;

export function toWebPlan(
  core: CorePlan,
  opts: { planId?: string; engine?: WebPlan['engine']; constraints?: OptimizeConstraints } = {},
): WebPlan {
  const rotations = core.placements.filter((p) => p.rot_deg !== 0).length;
  const cycleS = core.placements.length * CYCLE_BASE_S_PER_BOX + rotations * CYCLE_ROT_PENALTY_S;

  const metrics: PlanMetrics = {
    num_boxes: core.placements.length,
    unique_skus: new Set(core.placements.map((p) => p.sku_id)).size,
    num_layers: core.num_layers,
    volume_density: core.volume_density,
    density_uplift_pct: core.density_uplift_pct,
    stability_score: core.stability_score,
    support_score: core.support_score,
    com_score: core.com_score,
    total_weight_kg: core.total_weight_kg,
    stack_height_mm: core.stack_height_mm,
    est_build_time_min: Math.round((cycleS / 60) * 10) / 10,
    est_robot_cycle_s: Math.round(cycleS * 10) / 10,
    rotations_90: rotations,
  };

  const validation: ValidationReport = {
    is_valid: core.is_valid,
    stability_pass: core.stability_score >= 0.6,
    recommendations: core.recommendations,
  };

  return {
    plan_id: opts.planId ?? `plan_${Date.now()}`,
    created_at: new Date().toISOString(),
    engine: opts.engine ?? 'ts-port',
    constraints: opts.constraints,
    metrics,
    validation_report: validation,
    boxes: core.placements,
    unplaced: core.unplaced,
  };
}

export function planFromBoxes(
  boxes: BoxSpec[],
  constraints: OptimizeConstraints = {},
  pallet?: Partial<PalletSpec>,
  planId?: string,
): WebPlan {
  return toWebPlan(optimizePallet(boxes, pallet, constraints), { planId, constraints });
}
